STP: initial work

logicog opened the pull request:
This adds initial support for the (R/M)STP protocols. It adds commands to enable and disable STP In the enabled state, incoming STP packets are parsed and the root-bridge state is updated. Additionally, every 2 seconds HELLO packets are sent which contain the current belief of the STP.

A lot of the code for a proper STP implementation is missing such as all the state machines. The purpose of this early PR is therefore more to show how incoming packets can be handled with what UIP calls a device driver and that incoming and outgoing packets have a different frame formats due to RTL837x adding and requiring different kinds of tagging. Also shown is how to send packets only on 1 port, which is needed because STP-CNF packages refer to the port they have been sent on.

Note that the central tick timer's frequency seems to be off. This PR only handles the current situation, it does not try to fix it.

vDorst commented on rtlplayground.c:
```diff
@@ -54,6 +54,10 @@ __xdata uint8_t isRTL8373;
 volatile __xdata uint32_t ticks;
 volatile __xdata uint8_t sec_counter;
 volatile __xdata uint16_t sleep_ticks;
+__xdata uint8_t sfp_clock;
```
seems a typo.
sfp_clock. I think you mean stp_clock?
So second letter should be a `t` instead of a `f`.

vDorst commented on rtl837x_stp.c:
```diff
@@ -103,4 +197,55 @@ void stp_cnf_send(uint8_t port) __banked
+{
+	print_string("Enabling STP: ");
+	// Read STP port states: 00 disable, 01 blocking, 10 learning, 11 forwarding
+	reg_read_m(RTL837X_MSTP_STATES);
```
Do you need to read the register? If so, maybe add a note about it.
Because line below you clear all the data.
Same in `stp_off()`.

vDorst commented on rtl837x_stp.c:
```diff
@@ -103,4 +197,54 @@ void stp_cnf_send(uint8_t port) __banked
+void stp_setup(void) __banked
+{
+	print_string("Enabling STP: ");
+	// Read STP port states: 00 disable, 01 blocking, 10 learning, 11 forwarding
```
the comment `// Read ...` has less of a meaning because `reg_read_m()` is removed.
Same also happens in `stp_off()`.

logicog commented:
Ouch, should have tested a bit more thoroughly. I moved the comments and they should now be more meaningful. Also fixed the forgotten use of the bit_mask.

logicog commented:
Fixed the appearance of SFP_ instead of STP_.
